crates/red_knot_python_semantic/src/semantic_index/use_def.rs

```diff
@@ -544,7 +539,7 @@ impl<'db> UseDefMapBuilder<'db> {
             binding,
             SymbolDefinitions::Declarations(symbol_state.declarations().clone()),
         );
-        symbol_state.record_binding(def_id);
+        symbol_state.record_binding(def_id, self.scope_start_visibility);
```

Comment:
New bindings should only be considered visible if/when the flow is reachable

Reply:
> Is this the fix to the "visibility constraints only apply to prior, not subsequent, bindings" issue that @sharkdp pointed out on the very first terminals PR that used visibility constraints?

It is!

> Do we already have (and if not, can we add?) a test showing that a definition that occurs in the unreachable code after a terminal statement in a branch doesn't then become visible after that branch merges?

The `"unreachable"` assignment in the `try`/`else` clause should cover this, but I can add a more direct test of this as well.